### PR TITLE
issue 454, add support for AWS AutoScalingGroups

### PR DESCRIPTION
**Pull request by pachico**
~~**DO NOT MERGE, THIS REQUIRES FURTHER REVIEW AND AGREEMENT**~~

## Problem

This solves the feature request https://github.com/tailwarden/komiser/issues/454

## Solution

Implements the functionality to discover AutoScalingGroups

## Changes Made

- Added a struct that discovers ASGs.
- It requires a struct that satisfies the newly created AutoScalingGroup interface. This allows unit-testing it if not all (I don't think it's possible to achieve more with the current design) at least the meat of the functionality
- I also added a mock of such an interface using vanilla mockery 

## How to Test

From what I see in the mod file, the application has been written in Go 1.18, while I currently work with 1.20.
All CI operations should be easily reproducible and should not depend on local versions.

To execute the test, please run:

```bash
docker run --rm \
	--volume $(PWD):/app \
	--workdir /app \
	golang:1.18-alpine \
	ash -c "go mod tidy && CGO_ENABLED=0 go test ./providers/... -cover"
```

I would extend the same practice to `vet`, `fmt`, and all the tests that need to be executed before the build process.

The tests, which only cover one file, are successful.

```bash
ok      github.com/tailwarden/komiser/providers/aws/ec2 0.008s  coverage: 8.4% of statements
```

I also see many lines reporting other errors which I don't know where they come from.
Is this a known issue?

```bash
 github.com/tailwarden/komiser/providers/aws/cloudfront
providers/aws/cloudfront/distributions.go:54:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
providers/aws/cloudfront/distributions.go:80:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
providers/aws/cloudfront/distributions.go:106:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/aws/lambda
providers/aws/lambda/functions.go:49:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
providers/aws/lambda/functions.go:74:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/aws/apigateway
providers/aws/apigateway/apis.go:57:4: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/aws/sns
providers/aws/sns/topics.go:79:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/aws/sqs
providers/aws/sqs/queues.go:54:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
providers/aws/sqs/queues.go:80:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
providers/aws/sqs/queues.go:106:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/aws/s3
providers/aws/s3/buckets.go:56:4: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/digitalocean/k8s
providers/digitalocean/k8s/kubernetes.go:49:16: fmt.Sprintf format %d has arg kubernetesCluster.ID of wrong type string
# github.com/tailwarden/komiser/providers/digitalocean/network
providers/digitalocean/network/vpcs.go:19:3: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/digitalocean/kubernetes
providers/digitalocean/kubernetes/kubernetes.go:44:16: fmt.Sprintf format %d has arg kubernetesCluster.ID of wrong type string
providers/digitalocean/kubernetes/kubernetes.go:49:16: fmt.Sprintf format %d has arg kubernetesCluster.ID of wrong type string
# github.com/tailwarden/komiser/providers/linode/compute
providers/linode/compute/instances.go:47:16: fmt.Sprintf format %s has arg instance.ID of wrong type int
providers/linode/compute/lke_clusters.go:47:16: fmt.Sprintf format %s has arg lkeCluster.ID of wrong type int
# github.com/tailwarden/komiser/providers/linode/storage
providers/linode/storage/databases.go:30:16: fmt.Sprintf format %s has arg database.ID of wrong type int
providers/linode/storage/volumes.go:47:16: fmt.Sprintf format %s has arg volume.ID of wrong type int
# github.com/tailwarden/komiser/providers/linode/networking
providers/linode/networking/firewalls.go:47:16: fmt.Sprintf format %s has arg firewall.ID of wrong type int
providers/linode/networking/node_balancers.go:47:16: fmt.Sprintf format %s has arg nodeBalancer.ID of wrong type int
# github.com/tailwarden/komiser/providers/aws
providers/aws/aws.go:74:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
ok      github.com/tailwarden/komiser/providers/aws/ec2 0.008s  coverage: 8.4% of statements
```

## Notes

Pending

## Checklist

- [ ] Pending

## Reviewers

@[username of the reviewer]

### CODE REVIEW

File: providers/aws/ec2/autoscaling_groups_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/aws/aws-sdk-go-v2/service/autoscaling/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/tailwarden/komiser/mocks"
+	"github.com/tailwarden/komiser/models"
+	"github.com/tailwarden/komiser/providers/aws/ec2"
+)
+
+func TestDiscoverReturnsNoErrorIfHappyPath(t *testing.T) {
+
+	// Arrange
```
Comment (pachico) [inline providers/aws/ec2/autoscaling_groups_test.go:20]: I am quite a fan of the AAA pattern and I like it to be explicit in my code. However, I don't have strong opinions if you decide to remove the indicators for Arrange, Act, and Assert.

File: go.mod [hunk 2]
```diff
@@ -48,6 +49,7 @@ require (
 	github.com/segmentio/analytics-go v3.1.0+incompatible
 	github.com/sirupsen/logrus v1.9.0
 	github.com/siruspen/logrus v1.7.1
+	github.com/slack-go/slack v0.12.1
```
Comment (pachico) [inline go.mod:52]: You can ignore this change. My IDE simply moved it for reasons that are beyond my comprehension.

File: providers/aws/ec2/autoscaling_groups_client.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package ec2
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go-v2/service/autoscaling"
+)
+
+type AutoScalingGroupClient interface {
```
Comment (pachico) [inline providers/aws/ec2/autoscaling_groups_client.go:9]: This is the client that needs to be implemented to allow for easy unit testing.

File: providers/aws/ec2/autoscaling_groups.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/autoscaling"
+	. "github.com/tailwarden/komiser/models"
+	. "github.com/tailwarden/komiser/providers"
+)
+
+// This is to comply with current architecture as described in CONTRIBUTING.md
```
Comment (pachico) [inline providers/aws/ec2/autoscaling_groups.go:16]: The changes made to the relationship between `AutoScalingGroups` and the `Discover()` method represent the minimum necessary for efficient testing of the resource fetchers, without requiring a complete redesign.

That being said, I can't think of a reason why the `ASGDiscoverer` struct shouldn't implement a `Discoverer` interface that ALL discoverers implement. This way, the orchestration of the fetchers could also be efficiently tested.

File: providers/aws/ec2/autoscaling_groups_test.go (hunk 1 above)
Comment (mlabouardy) [reply providers/aws/ec2/autoscaling_groups_test.go:20]: I'm a big fan of AAA as always, so all good :)

**Review by mlabouardy — APPROVED** [review]
@pachico I've made the following changes to your PR:
- Fixed the errors reported while running tests
- Triggered ASG fetching workflow
- Triggered tests within CI pipeline
If all good for you, I can merge this one 🚀

**Comment by pachico** [comment]
> @pachico I've made the following changes to your PR:
> 
> * Fixed the errors reported while running tests
> * Triggered ASG fetching workflow
> * Triggered tests within CI pipeline
>   If all good for you, I can merge this one 🚀

Thanks a lot.
I want to do some minor tweaks (remove unnecessary comments, avoid magic strings) and test it with real cases and then I'll come back to you again for the final approval, if that's ok.

Cheers

**Comment by mlabouardy** [comment]
> > @pachico I've made the following changes to your PR:
> > 
> > * Fixed the errors reported while running tests
> > * Triggered ASG fetching workflow
> > * Triggered tests within CI pipeline
> >   If all good for you, I can merge this one 🚀
> 
> Thanks a lot. I want to do some minor tweaks (remove unnecessary comments, avoid magic strings) and test it with real cases and then I'll come back to you again for the final approval, if that's ok.
> 
> Cheers

Definitely! Thanks again for making the code better :)

**Comment by mlabouardy** [comment]
> > @pachico I've made the following changes to your PR:
> > 
> > * Fixed the errors reported while running tests
> > * Triggered ASG fetching workflow
> > * Triggered tests within CI pipeline
> >   If all good for you, I can merge this one 🚀
> 
> Thanks a lot. I want to do some minor tweaks (remove unnecessary comments, avoid magic strings) and test it with real cases and then I'll come back to you again for the final approval, if that's ok.
> 
> Cheers

Hey @pachico if those changes are not blocking, I could merge this one to have the ASG support in this weekly release and you can open another PR with the minor tweaks, wdyt?

**Comment by pachico** [comment]
I think I will be able to unblock it within the next 2 hours.
Would that be suitable to you?

**Comment by mlabouardy** [comment]
> I think I will be able to unblock it within the next 2 hours. Would that be suitable to you?

Yes perfect

**Pull request merged**